cpp/server/proxy.cc

```diff
@@ -0,0 +1,134 @@
+#include "server/proxy.h"
+
+#include <event2/http.h>
+#include <evhttp.h>
+
+#include "monitoring/monitoring.h"
+#include "net/url_fetcher.h"
+#include "proto/ct.pb.h"
+#include "server/json_output.h"
+#include "util/libevent_wrapper.h"
+
+
+namespace cert_trans {
+
+namespace {
+static Counter<std::string>* total_proxied_requests(
+    Counter<std::string>::New("total_proxied_requests", "path",
+                              "Number of proxied API requests by path."));
+static Counter<std::string, int>* total_proxied_responses(
+    Counter<std::string, int>::New("total_proxied_responses", "path",
+                                   "status_code",
+                                   "Number of proxied API requests by path "
+                                   "and status code."));
+
+
+void ProxyRequestDone(libevent::Base* base, JsonOutput* output,
+                      evhttp_request* request, const std::string& path,
+                      UrlFetcher::Response* response, util::Task* task) {
+  CHECK_NOTNULL(request);
+  CHECK_NOTNULL(response);
+  CHECK_NOTNULL(task);
+  total_proxied_requests->Increment(path);
+  total_proxied_responses->Increment(path, response->status_code);
+
+  if (!task->status().ok()) {
+    return output->SendError(request, HTTP_INTERNAL,
+                             "Proxied request failed.");
+  }
+
+  // TODO(alcutter): Consider retrying the proxied request some number of times
+  // in the case where the request fails.
+  for (auto it(response->headers.begin()); it != response->headers.end();
+       ++it) {
+    CHECK_EQ(evhttp_add_header(evhttp_request_get_output_headers(request),
+                               it->first.c_str(), it->second.c_str()),
+             0);
+  }
+  CHECK_GT(evbuffer_add_printf(evhttp_request_get_output_buffer(request), "%s",
+                               response->body.c_str()),
```

Comment:
This is just a suggestion for avoiding a copy, but if you want to keep it simple, make the copy and delete ```response``` here, that's fine by me as well.